_posts/2015-06-25-infrastructure-as-code.md

```diff
@@ -0,0 +1,82 @@
+
+{% image blog-jenkins-polyglot.png %}{% endimage %}
+
+# Jenkins configuration
+
+Jenkins makes it very easy to introduce new functionality by installing one of the [many available plugins](https://wiki.jenkins-ci.org/display/JENKINS/Plugins).
+Unfortunately not every plugin turns out to work as expected or in the worst case can potentially bring down the whole Jenkins instance.
+To evaluate unknown plugins by installing them into your production CI infrastructure bears problems, since they often leave unwanted traces behind, even after uninstallation.
+Even making changes to a job configuration without adding any new plugins can lead to broken build jobs, negatively affecting the productivity of whole teams.
+
+## The solution
+
+We were looking for ways to make it safe and easy to improve our build infrastructure.
+Using version control like [GitHub](https://github.com/) for all of our source code, it was natural to treat our infrastructure the same way and be able to roll back to a previous version without any hassle.
+Versioning whole images of Jenkins virtual servers didn't seem feasible and plugins like [JobConfigHistory](https://wiki.jenkins-ci.org/display/JENKINS/JobConfigHistory+Plugin) didn't go far enough.
+
+## Server provisioning using Ansible and Vagrant
+
+We started by automating the installation of Jenkins, including all required plugins.
+The idea is to set up a Jenkins server from scratch by running just a single command, while still being able to use the same facilities to update the installation, e.g. with new plugins, using the same technology.
+To achieve this, we decided to use [Ansible](http://www.ansible.com/home), mainly due to existing know-how.
+
+Once the basic installation of Jenkins and its prerequisites is done, the interesting part begins: installing and configuring plugins.
+The installation of additional plugins is done by utilising the [Jenkins Command Line Interpreter (CLI)](https://wiki.jenkins-ci.org/display/JENKINS/Jenkins+CLI).
+This allows us to provide plugin names and explicit versions to be installed.
+The tricky part is getting the global configuration of Jenkins right.
+The pragmatic solution we chose was to create the `config.xml` locally (but templated) and copy it over to Jenkins using the according Ansible task.
+
+With the toolset described so far, the basic setup can be easily recreated any time.
+And since this is so easy now, we usually test our setup scripts by installing them to local [Vagrant](https://www.vagrantup.com/) machines.
+Especially for the global Jenkins configuration this is an essential step, which is even easier and faster due to Ansible's smart way of tracking installation state.
+Just reapply the Ansible playbook, and changes in configuration are active - and you are sure they are based on your version-managed infrastructure code!
```

Comment:
doesn't flow so nicely, although i do not understand the full context. Proposal:

Just reapply the Ansible playbook and changes in the configuration are automatically /"integrated/ incorporated/ built-in"/; then you can be sure, that they are based on your version-managed infrastructure code!